src/lib/Libattr/attr_func.c

```diff
@@ -1334,3 +1335,370 @@ strtok_quoted(char *source, const char *delimiters)
 	prune_quotes(stok);
 	return (stok);
 }
+
+/*
+ *
+ *	attropl2attrl - convert an attropl struct to an attrl struct
+ *
+ *	  from - the attropl struct to convert
+ *
+ *	returns newly converted attrl struct
+ *
+ */
+struct attrl *
+attropl2attrl(struct attropl *from)
+{
+	struct attrl *ap = NULL, *rattrl = NULL;
+
+	while (from != NULL) {
+		if (ap == NULL) {
+			if ((ap = (struct attrl*) malloc(sizeof(struct attrl))) == NULL) {
+				perror("Out of memory");
+				return NULL;
+			}
+			rattrl = ap;
+		}
+		else  {
+			if ((ap->next = (struct attrl*) malloc(sizeof(struct attrl))) == NULL) {
+				perror("Out of memory");
+				return NULL;
+			}
+			ap = ap->next;
+		}
+
+		ap->name = NULL;
+		ap->resource = NULL;
+		ap->value = NULL;
+		ap->next = NULL;
+		if (from->name != NULL) {
+			if ((ap->name = (char*) malloc(strlen(from->name) + 1)) == NULL) {
+				perror("Out of memory");
+				free_attrl(ap);
+				return NULL;
+			}
+			strcpy(ap->name, from->name);
+		}
+		if (from->resource != NULL) {
+			if ((ap->resource = (char*) malloc(strlen(from->resource) + 1)) == NULL) {
+				perror("Out of memory");
+				free_attrl(ap);
+				return NULL;
+			}
+			strcpy(ap->resource, from->resource);
+		}
+		if (from->value != NULL) {
+			if ((ap->value = (char*) malloc(strlen(from->value) + 1)) == NULL) {
+				perror("Out of memory");
+				free_attrl(ap);
+				return NULL;
+			}
+			strcpy(ap->value, from->value);
+		}
+		from = from->next;
+	}
+
+	return rattrl;
+}
+
+/**
+ *  @brief attrl copy constructor
+ *
+ *  @param[in] oattr - attrl to dup
+ *
+ *  @return dup'd attrl
+ */
+
+struct attrl *
+dup_attrl(struct attrl *oattr)
+{
+	struct attrl *nattr;
+
+	if (oattr == NULL)
+		return NULL;
+
+	nattr = new_attrl();
+	if (nattr == NULL)
+		return NULL;
+	if (oattr->name != NULL)
+		nattr->name = strdup(oattr->name);
+	if (oattr->resource != NULL)
+		nattr->resource = strdup(oattr->resource);
+	if (oattr->value != NULL)
+		nattr->value = strdup(oattr->value);
+
+	nattr->op = oattr->op;
+	return nattr;
+}
+
+/**
+ * @brief copy constructor for attrl list
+ * @param oattr_list - list to dup
+ * @return dup'd attrl list
+ */
+
+struct attrl *
+dup_attrl_list(struct attrl *oattr_list)
+{
+	struct attrl *nattr_head = NULL;
+	struct attrl *nattr;
+	struct attrl *nattr_prev = NULL;
+	struct attrl *oattr;
+
+	if (oattr_list == NULL)
+		return NULL;
+
+	for (oattr = oattr_list; oattr != NULL; oattr = oattr->next) {
+		nattr = dup_attrl(oattr);
+		if (nattr_prev == NULL) {
+			nattr_head = nattr;
+			nattr_prev = nattr_head;
+		} else {
+			nattr_prev->next = nattr;
+			nattr_prev = nattr;
+		}
+	}
+	return nattr_head;
+}
+
+/**
+ *	@brief create a new attrl structure and initialize it
+ */
+struct attrl *
+new_attrl()
```

Comment:
I substituted the mallocs with calls to new_attrl() at all places, thanks for pointing it out.